token-swap/program/src/curve.rs

```diff
@@ -0,0 +1,170 @@
+}
+
+impl SwapResult {
+    /// SwapResult for swap from one currency into another, given pool information
+    /// and fee
+    pub fn swap_to(
+        source: u64,
+        source_amount: u64,
+        dest_amount: u64,
+        fee_numerator: u64,
+        fee_denominator: u64,
+    ) -> Option<SwapResult> {
+        let invariant = source_amount.checked_mul(dest_amount)?;
+        let new_source = source_amount.checked_add(source)?;
+        let new_destination = invariant.checked_div(new_source)?;
+        let remove = dest_amount.checked_sub(new_destination)?;
+        let fee = remove
+            .checked_mul(fee_numerator)?
+            .checked_div(fee_denominator)?;
+        let new_destination = new_destination.checked_add(fee)?;
+        let amount_swapped = remove.checked_sub(fee)?;
+        Some(SwapResult {
+            new_source,
+            new_destination,
+            amount_swapped,
+        })
+    }
+}
+
+/// The Uniswap invariant calculator.
+pub struct ConstantProduct {
+    /// Token A
+    pub token_a: u64,
+    /// Token B
+    pub token_b: u64,
+    /// Fee numerator
+    pub fee_numerator: u64,
+    /// Fee denominator
+    pub fee_denominator: u64,
+}
+
+impl ConstantProduct {
+    /// Swap token a to b
+    pub fn swap_a_to_b(&mut self, token_a: u64) -> Option<u64> {
+        let result = SwapResult::swap_to(
+            token_a,
+            self.token_a,
+            self.token_b,
+            self.fee_numerator,
+            self.fee_denominator,
+        )?;
+        self.token_a = result.new_source;
+        self.token_b = result.new_destination;
+        Some(result.amount_swapped)
+    }
+
+    /// Swap token b to a
+    pub fn swap_b_to_a(&mut self, token_b: u64) -> Option<u64> {
+        let result = SwapResult::swap_to(
+            token_b,
+            self.token_b,
+            self.token_a,
+            self.fee_numerator,
+            self.fee_denominator,
+        )?;
+        self.token_b = result.new_source;
+        self.token_a = result.new_destination;
+        Some(result.amount_swapped)
+    }
+}
+
+/// Conversions for pool tokens, how much to deposit / withdraw, along with
+/// proper initialization
+pub struct PoolTokenConverter {
+    /// Total supply
+    pub supply: u64,
+    /// Token A amount
+    pub token_a: u64,
+    /// Token B amount
+    pub token_b: u64,
+}
+
+impl PoolTokenConverter {
+    /// Create a converter based on existing market information
+    pub fn new_existing(supply: u64, token_a: u64, token_b: u64) -> Self {
+        Self {
+            supply,
+            token_a,
+            token_b,
+        }
+    }
+
+    /// Create a converter for a new pool token, no supply present yet.
+    /// According to Uniswap, the geometric mean protects the pool creator
+    /// in case the initial ratio is off the market.
+    pub fn new_pool(token_a: u64, token_b: u64) -> Option<Self> {
+        let supply = math::geometric_mean(&[token_a, token_b])?;
```

Comment:
That was my first consideration, but I was worried it would cause a problem in the VM.  If they work, I'll definitely use them under the hood.  Since this is a geometric mean it will stay within the size of the inputs (u64 in this case).  The problem is that this will fail for `geometric_mean(&[u64::MAX, u64::MAX, u64::MAX])` even with u128 under the hood.  I took a look at how Balancer was doing things, and they're just hardcoding `100 * 10 ^18` for the initial supply.  I'm considering more and more that we might be best served with a fixed starting point and nixing this math.  Since `u64::MAX` is roughly `10 ^ 19`, perhaps something like `10 ^ 9` would be good?

Reply:
I definitely like the simplicity of a fixed starting point, and `10^9` seems about right.
I've been looking for some Balancer documentation around their decision for initial supply; is there anything you can point me at?

Reply:
I literally dug into the source code :smile: https://github.com/balancer-labs/balancer-core/blob/140df49361a58e6c79b395964be98387702a7c0d/contracts/BConst.sol#L33

Reply:
I can definitely go to the fixed number -- could any of this math be useful elsewhere?

Reply:
I don't know of another need off the top of my head, no reason to keep. But if you see anything coming down the pipe wrt balancer, feel free to leave it!